testing/describes.js

```diff
@@ -0,0 +1,387 @@
+const SUB = ' ';
+
+
+/** @type {number} */
+let iframeCount = 0;
+
+
+/**
+ * @typedef {{
+ *   fakeClock: (boolean|undefined),
+ *   fakeRegisterElement: (boolean|undefined),
+ * }}
+ */
+export let TestSpec;
+
+
+/**
+ * - ampdoc: "single", "shadow", "multi", "none".
+ *
+ * @typedef {{
+ *   runtimeOn: (boolean|undefined),
+ *   ampdoc: (string),
+ *   params: (!Object<string, string>|undefined),
+ * }}
+ */
+export let AmpTestSpec;
+
+
+/**
+ * @typedef {{
+ *   win: !Window,
+ *   extensions: !Extensions,
+ *   ampdocService: !AmpDocService,
+ *   ampdoc: (!AmpDoc|undefined),
+ *   flushVsync: function(),
+ * }}
+ */
+export let AmpTestEnv;
+
+
+/**
+ * A test with a sandbox.
+ * @param {string} name
+ * @param {!TestSpec} spec
+ * @param {function()} fn
+ */
+export function sandboxed(name, spec, fn) {
+  return describeEnv(name, fn, [new SandboxFixture(spec)]);
+}
+
+
+/**
+ * A test with a fake window.
+ * @param {string} name
+ * @param {{
+ *   win: !FakeWindowSpec,
+ *   amp: (!AmpTestSpec|undefined),
+ * }} spec
+ * @param {function({
+ *   win: !FakeWindow,
+ *   amp: (!AmpTestEnv|undefined),
+ * })} fn
+ */
+export function fakeWin(name, spec, fn) {
+  return describeEnv(name, fn, [
+      new SandboxFixture(spec),
+      new FakeWinFixture(spec),
+      new AmpFixture(spec),
+    ]);
+}
+
+
+/**
+ * A test with a real (iframed) window.
+ * @param {string} name
+ * @param {{
+ *   fakeRegisterElement: (boolean|undefined),
+ *   amp: (!AmpTestSpec|undefined),
+ * }} spec
+ * @param {function({
+ *   win: !Window,
+ *   iframe: !HTMLIFrameElement,
+ *   amp: (!AmpTestEnv|undefined),
+ * })} fn
+ */
+export function realWin(name, spec, fn) {
+  return describeEnv(name, fn, [
+      new SandboxFixture(spec),
+      new RealWinFixture(spec),
+      new AmpFixture(spec),
+    ]);
+}
+
+
+/**
+ * A test with in a described environment.
+ * @param {string} name
+ * @param {function(!Object)} fn
+ * @param {!Array<?Fixture>} fixtures
+ */
+function describeEnv(name, fn, fixtures) {
+  return describe(name, function() {
+
+    const env = Object.create(null);
+
+    beforeEach(() => {
+      let totalPromise = undefined;
+      // Set up all fixtures.
+      fixtures.forEach((fixture, index) => {
+        if (!fixture || !fixture.isOn()) {
+          return;
+        }
+        if (totalPromise) {
+          totalPromise = totalPromise.then(() => fixture.setup(env));
+        } else {
+          const res = fixture.setup(env);
+          if (res && typeof res.then == 'function') {
+            totalPromise = res;
+          }
+        }
+      });
+      return totalPromise;
+    });
+
+    afterEach(() => {
+      // Tear down all fixtures.
+      fixtures.forEach(fixture => {
+        if (!fixture || !fixture.isOn()) {
+          return;
+        }
+        fixture.teardown(env);
+      });
+
+      // Delete all other keys.
+      for (const key in env) {
+        delete env[key];
+      }
+    });
+
+    describe(SUB, function() {
```

Comment:
So that the callback we call (`fn`) can call its own `beforeEach` and `afterEach` and we observe the nested behavior.